pkg/controller/stacks/stack/stack_test.go

```diff
@@ -1688,6 +1701,657 @@ func Test_stackHandler_prepareHostAwarePodSpec(t *testing.T) {
+				ConversionReviewVersions: nil,
+			},
+			PreserveUnknownFields: &t,
+		},
+	}
+	for _, m := range cm {
+		m(&c)
+	}
+	return c
+}
+
+type crModifier func(*rbac.ClusterRole)
+
+func withClusterRoleLabels(labels map[string]string) crModifier {
+	return func(cr *rbac.ClusterRole) {
+		meta.AddLabels(cr, labels)
+	}
+}
+
+func withClusterRoleRules(rules []rbac.PolicyRule) crModifier {
+	return func(cr *rbac.ClusterRole) {
+		cr.Rules = append(cr.Rules, rules...)
+
+	}
+}
+
+func clusterRole(name string, crm ...crModifier) rbac.ClusterRole {
+	c := rbac.ClusterRole{}
+	c.SetName(name)
+	for _, m := range crm {
+		m(&c)
+	}
+	return c
+}
+
+func Test_crdListFulfilled(t *testing.T) {
+	type args struct {
+		want v1alpha1.CRDList
+		got  []apiextensionsv1beta1.CustomResourceDefinition
+	}
+
+	const (
+		group      = "samples.upbound.io"
+		version    = "v1alpha1"
+		kind       = "Mytype"
+		plural     = "mytypes"
+		apiVersion = group + "/" + version
+	)
+
+	tests := []struct {
+		name    string
+		args    args
+		wantErr error
+	}{
+		{
+			name:    "MissingFromCRDList",
+			args:    args{v1alpha1.CRDList{metav1.TypeMeta{Kind: kind, APIVersion: apiVersion}}, []apiextensionsv1beta1.CustomResourceDefinition{}},
+			wantErr: fmt.Errorf(`Missing CRD with APIVersion %q and Kind %q`, apiVersion, kind),
+		},
+		{
+			name:    "WrongCRDVersion",
+			args:    args{v1alpha1.CRDList{metav1.TypeMeta{Kind: kind, APIVersion: apiVersion}}, []apiextensionsv1beta1.CustomResourceDefinition{crd(withCRDGroupKind(group, kind), withCRDVersion("foo"))}},
+			wantErr: fmt.Errorf(`Missing CRD with APIVersion %q and Kind %q`, apiVersion, kind),
+		},
+		{
+			name:    "DifferentCRDKind",
+			args:    args{v1alpha1.CRDList{metav1.TypeMeta{Kind: kind, APIVersion: apiVersion}}, []apiextensionsv1beta1.CustomResourceDefinition{crd(withCRDGroupKind(group, "foo"), withCRDVersion(version))}},
+			wantErr: fmt.Errorf(`Missing CRD with APIVersion %q and Kind %q`, apiVersion, kind),
+		},
+		{
+			name:    "PartialMatch",
+			args:    args{v1alpha1.CRDList{metav1.TypeMeta{Kind: kind, APIVersion: apiVersion}, metav1.TypeMeta{Kind: kind + "z", APIVersion: apiVersion}}, []apiextensionsv1beta1.CustomResourceDefinition{crd(withCRDGroupKind(group, kind), withCRDVersion(version))}},
+			wantErr: fmt.Errorf(`Missing CRD with APIVersion %q and Kind %q`, apiVersion, kind+"z"),
+		},
+		{
+			name:    "Success",
+			args:    args{v1alpha1.CRDList{metav1.TypeMeta{Kind: kind, APIVersion: apiVersion}}, []apiextensionsv1beta1.CustomResourceDefinition{crd(withCRDGroupKind(group, kind), withCRDVersion(version))}},
+			wantErr: nil,
+		},
+		{
+			name:    "SuccessMultiple",
+			args:    args{v1alpha1.CRDList{metav1.TypeMeta{Kind: kind, APIVersion: apiVersion}, metav1.TypeMeta{Kind: kind + "z", APIVersion: apiVersion}}, []apiextensionsv1beta1.CustomResourceDefinition{crd(withCRDGroupKind(group, kind), withCRDVersion(version)), crd(withCRDGroupKind(group, kind+"z"), withCRDVersion(version))}},
```

Comment:
consider making these really long lines to be multi-line for easier readability, it's hard to tell what objects are involved in each test case

Reply:
I may have wrapped some lines that were not in my PR in an attempt to clean this up visually.